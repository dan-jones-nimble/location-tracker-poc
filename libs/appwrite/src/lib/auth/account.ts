import { Account } from 'appwrite';
import { client } from '../client';

export const account = new Account(client);

export const getAccount = async () => {
  try {
    return await account.get();
  } catch (e: any) {
    console.error(e.message);
    return false;
  }
};

export const logout = async () => {
  try {
    console.log('account: ', await account.get());
    const sessions = await account.listSessions();
    console.log('session: ', sessions);
    return true;
  } catch (e: any) {
    console.error(e.message);
    return false;
  }
};
