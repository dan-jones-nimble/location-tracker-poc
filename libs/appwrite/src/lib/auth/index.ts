export * from './account';
export * from './loginEmailSession';
