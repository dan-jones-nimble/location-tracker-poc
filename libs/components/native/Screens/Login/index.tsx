// Import
import { useState } from 'react';
import { Button, StyleSheet, TextInput, View } from 'react-native';

// Context
import { useLoggedIn } from '@nx-expo/context';

// Appwrite
import { emailLogin, ICredentials } from '@nx-expo/appwrite';

export const LoginScreen = () => {
  const { logIn } = useLoggedIn();
  const [credentials, setCredentials] = useState<ICredentials>({
    email: '',
    password: ''
  });

  const handleInput = (name: string, value: string) => {
    setCredentials((prevState) => ({
      ...prevState,
      [name]: value
    }));
  };

  const submitCredentials = async () => {
    const attemptedEmailLogin = await emailLogin({
      email: credentials.email,
      password: credentials.password
    });
    if (attemptedEmailLogin?.$id) logIn();
  };

  return (
    <View style={styles.container}>
      <View>
        <TextInput
          style={styles.textInput}
          onChange={(e) => handleInput('email', e.nativeEvent.text)}
          autoComplete="email"
          autoFocus
          textAlign="center"
        />
        <TextInput
          style={styles.textInput}
          onChange={(e) => handleInput('password', e.nativeEvent.text)}
          aria-hidden
          autoComplete="current-password"
          textAlign="center"
        />
      </View>
      <Button title="Log In" onPress={submitCredentials} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    gap: 16
  },
  textInput: {
    display: 'flex',
    fontSize: 20,
    minWidth: '60%',
    borderColor: '#a7a7a7ee',
    borderBottomWidth: 2,
    borderRadius: 4,
    padding: 4
  }
});
