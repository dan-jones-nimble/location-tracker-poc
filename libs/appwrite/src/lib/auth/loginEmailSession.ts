import { account } from './account';

interface ICredentials {
  email: string;
  password: string;
}

export const emailLogin = async (credentials: ICredentials) => {
  return await account.createEmailSession(
    credentials.email,
    credentials.password
  );
};
