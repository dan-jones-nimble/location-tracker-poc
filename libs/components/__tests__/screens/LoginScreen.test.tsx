import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { ICredentials } from "../../../appwrite/src/lib/auth/loginEmailSession";
import { LoginScreen } from "../../native/Screens/Login";
import { Button } from "react-native";
import React from "react";

const mockEmailLogin = jest.fn();
const mockLogin = jest.fn();
const mockWobbleButton = (props) => <Button {...props} />;

jest.mock('@nx-expo/appwrite', () => ({
    emailLogin: (args) => new Promise(() => mockEmailLogin(args)),
}));

jest.mock("../../native/WobbleButton/index", () => ({
    WobbleButton: (props) => mockWobbleButton(props),
}));

jest.mock("@nx-expo/context", () => ({
    useLoggedIn: () => ({ logIn: mockLogin }),
}));

describe("Login Screen", () => {
    it("matches snapshot", () => {
        const tree = render(<LoginScreen />);
        expect(tree).toMatchSnapshot();
    });

    it("button pressed, emailLogin called with correct email and password", async () => {
        const creds: ICredentials = {
            email: "test@email.com",
            password: "password"
        };
        const tree = render(<LoginScreen />);
        const emailInput = tree.getByTestId("email");
        const passwordInput = tree.getByTestId("password");
        const button = tree.getByRole("button");
        fireEvent.changeText(emailInput, creds.email);
        fireEvent.changeText(passwordInput, creds.password);
        fireEvent.press(button);
        await waitFor(() => expect(mockEmailLogin).toBeCalledWith(creds));
    });
});