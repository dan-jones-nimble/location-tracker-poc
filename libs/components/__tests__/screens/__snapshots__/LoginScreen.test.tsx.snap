// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen matches snapshot 1`] = `
<View
  style={
    {
      "display": "flex",
      "gap": 16,
    }
  }
>
  <View>
    <TextInput
      autoCapitalize="none"
      autoComplete="email"
      autoCorrect={false}
      autoFocus={true}
      inputMode="email"
      keyboardType="email-address"
      onChangeText={[Function]}
      returnKeyType="next"
      style={
        {
          "borderBottomWidth": 2,
          "borderColor": "#a7a7a7ee",
          "borderRadius": 4,
          "display": "flex",
          "fontSize": 20,
          "minWidth": "60%",
          "padding": 4,
        }
      }
      testID="email"
      textAlign="center"
    />
    <TextInput
      aria-hidden={true}
      autoCapitalize="none"
      autoComplete="current-password"
      autoCorrect={false}
      onChangeText={[Function]}
      returnKeyType="done"
      secureTextEntry={true}
      style={
        {
          "borderBottomWidth": 2,
          "borderColor": "#a7a7a7ee",
          "borderRadius": 4,
          "display": "flex",
          "fontSize": 20,
          "minWidth": "60%",
          "padding": 4,
        }
      }
      testID="password"
      textAlign="center"
    />
  </View>
  <View
    collapsable={false}
    style={
      {
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Log In
        </Text>
      </View>
    </View>
  </View>
</View>
`;
